From b985f82532fc444064dccb2a587c08ee8f792cdb Mon Sep 17 00:00:00 2001
From: "K.Takata" <kentkt@csc.jp>
Date: Tue, 2 Oct 2018 16:52:32 +0900
Subject: [PATCH 13/27] nsis: Fix installer directory

Fix that the installer didn't check the old version of installed
directory and also $VIM.
Check $VIM first, then check the old directory from the registry, then
fallback to $PROGRAMFILES(X86)\Vim.
---
 nsis/gvim.nsi | 87 +++++++++++++++++++++++++++++----------------------
 1 file changed, 49 insertions(+), 38 deletions(-)

diff --git a/nsis/gvim.nsi b/nsis/gvim.nsi
index dace234b9..9a7a2c126 100644
--- a/nsis/gvim.nsi
+++ b/nsis/gvim.nsi
@@ -116,8 +116,8 @@ XPStyle on
 !define MUI_FINISHPAGE_RUN_TEXT            $(str_show_readme)
 !define MUI_FINISHPAGE_RUN_PARAMETERS      "-R $\"$0\README.txt$\""
 
-# This adds '\vim' to the user choice automagically.  The actual value is
-# obtained below with ReadINIStr.
+# This adds '\Vim' to the user choice automagically.  The actual value is
+# obtained below with CheckOldVim.
 InstallDir "$PROGRAMFILES\Vim"
 
 # Types of installs we can perform:
@@ -196,6 +196,34 @@ ReserveFile ${VIMSRC}\installw32.exe
 ##########################################################
 # Functions
 
+# Get parent directory
+# Share this function both on installer and uninstaller
+!macro GetParent un
+Function ${un}GetParent
+  Exch $0 ; old $0 is on top of stack
+  Push $1
+  Push $2
+  StrCpy $1 -1
+  ${Do}
+    StrCpy $2 $0 1 $1
+    ${If} $2 == ""
+    ${OrIf} $2 == "\"
+      ${ExitDo}
+    ${EndIf}
+    IntOp $1 $1 - 1
+  ${Loop}
+  StrCpy $0 $0 $1
+  Pop $2
+  Pop $1
+  Exch $0 ; put $0 on top of stack, restore $0 to original value
+FunctionEnd
+!macroend
+
+!insertmacro GetParent ""
+!insertmacro GetParent "un."
+
+# Check if Vim is already installed.
+# return: Installed directory. If not found, it will be empty.
 Function CheckOldVim
   Push $0
   Push $R0
@@ -207,7 +235,7 @@ Function CheckOldVim
   ${EndIf}
 
   ClearErrors
-  StrCpy $0 0	  # Found flag
+  StrCpy $0  ""   # Installed directory
   StrCpy $R0 0    # Sub-key index
   StrCpy $R1 ""   # Sub-key
   ${Do}
@@ -242,7 +270,11 @@ Function CheckOldVim
       ${Continue}
     ${EndIf}
 
-    StrCpy $0 1	  # Found
+    # Found
+    Push $R2
+    call GetParent
+    call GetParent
+    Pop $0   # Vim directory
     ${ExitDo}
 
   ${Loop}
@@ -254,9 +286,10 @@ Function CheckOldVim
   Pop $R2
   Pop $R1
   Pop $R0
-  Exch $0 ; put $0 on top of stack, restore $0 to original value
+  Exch $0  # put $0 on top of stack, restore $0 to original value
 FunctionEnd
 
+##########################################################
 Section "$(str_section_old_ver)" id_section_old_ver
 	SectionIn 1 2 3 RO
 
@@ -266,11 +299,13 @@ Section "$(str_section_old_ver)" id_section_old_ver
 	nsExec::Exec "$TEMP\install.exe -uninstall-check"
 	Pop $3
 	Delete $TEMP\install.exe
+	Delete $TEMP\vimini.ini   # install.exe creates this, but we don't need it.
 
 	# We may have been put to the background when uninstall did something.
 	BringToFront
 SectionEnd
 
+##########################################################
 Function .onInit
 #  MessageBox MB_YESNO|MB_ICONQUESTION \
 #	"This will install Vim ${VER_MAJOR}.${VER_MINOR} on your computer.$\n Continue?" \
@@ -279,32 +314,27 @@ Function .onInit
 #	    Abort ; causes installer to quit.
 #	NoAbort:
 
+  # Check $VIM
+  ReadEnvStr $INSTDIR "VIM"
+
   call CheckOldVim
   Pop $3
-  ${If} $3 == 0
+  ${If} $3 == ""
     # No old versions of Vim found. Unselect and hide the section.
     !insertmacro UnselectSection ${id_section_old_ver}
     SectionSetInstTypes ${id_section_old_ver} 0
     SectionSetText ${id_section_old_ver} ""
+  ${Else}
+    ${If} $INSTDIR == ""
+      StrCpy $INSTDIR $3
+    ${EndIf}
   ${EndIf}
 
-  # Install will have created a file for us that contains the directory where
-  # we should install.  This is $VIM if it's set.  This appears to be the only
-  # way to get the value of $VIM here!?
-  ReadINIStr $INSTDIR $TEMP\vimini.ini vimini dir
-  Delete $TEMP\vimini.ini
-
-  # If ReadINIStr failed or did not find a path: use the default dir.
+  # If did not find a path: use the default dir.
   ${If} $INSTDIR == ""
     StrCpy $INSTDIR "$PROGRAMFILES\Vim"
   ${EndIf}
 
-  # Should check for the value of $VIM and use it.  Unfortunately I don't know
-  # how to obtain the value of $VIM
-  # IfFileExists "$VIM" 0 No_Vim
-  #   StrCpy $INSTDIR "$VIM"
-  # No_Vim:
-
   # User variables:
   # $0 - holds the directory the executables are installed to
   # $1 - holds the parameters to be passed to install.exe.  Starts with OLE
@@ -350,25 +380,6 @@ FunctionEnd
 #  "Vim ${VER_MAJOR}.${VER_MINOR} has been (partly) removed from your system"
 #FunctionEnd
 
-Function un.GetParent
-  Exch $0 ; old $0 is on top of stack
-  Push $1
-  Push $2
-  StrCpy $1 -1
-  ${Do}
-    StrCpy $2 $0 1 $1
-    ${If} $2 == ""
-    ${OrIf} $2 == "\"
-      ${ExitDo}
-    ${EndIf}
-    IntOp $1 $1 - 1
-  ${Loop}
-  StrCpy $0 $0 $1
-  Pop $2
-  Pop $1
-  Exch $0 ; put $0 on top of stack, restore $0 to original value
-FunctionEnd
-
 ##########################################################
 Section "$(str_section_exe)" id_section_exe
 	SectionIn 1 2 3 RO
-- 
2.17.0

